crates/bevy_core/src/name.rs

```diff
@@ -229,9 +229,9 @@ mod tests {
         let e2 = world.spawn(name.clone()).id();
         let mut query = world.query::<NameOrEntity>();
         let d1 = query.get(&world, e1).unwrap();
-        let d2 = query.get(&world, e2).unwrap();
```

Comment:
This needing to be moved is suspicious. Can you not call get multiple times on an immutable borrow of the world anymore?

Reply:
Not in this case, but I promise it's not as bad as it looks!  

`QueryState::get()` takes `&mut self` so that it can call `update_archetypes()`.  And `NameOrEntity` uses `#[derive(QueryData)]`, which now generates an item struct that includes the `'s` lifetime.  So the item is treated as having a mutable borrow of the `QueryState`.  

This doesn't affect `Query::get()`, which always takes `&self`.  

And this doesn't affect concrete types where `QueryData::Item` doesn't include `'s`.  That is, it only affects generic parameters, `#[derive(QueryData)]` types, and (in the future) `FilteredEntityRef|Mut`.  If we use `let mut query = world.query::<(Option<&Name>, Entity)>();` here, it will compile.  

And there is a workaround in this case, which is to call `query.update_archetypes()` manually and then use `query.get_manual()`, which takes `&self` and so will allow overlapping borrows.  

But, yeah, it is a breaking change here.

Reply:
After digesting this for a bit, maybe it isn't such a bad change. `get` takes an `&mut self` so hopefully it won't be that surprising for new users if you can't get multiple items out of it. And the workaround is more performant, so leading users to that is probably better.

 But we should document the workaround on `QueryState::get` and in the migration guide.

Reply:
Does this not also affect query iteration? As in, for the affected types, any item would have to be dropped before another can be retrieved?
(Like what happens in `QueryManyIter::fetch_next()`, where item lifetimes are shrunk like this intentionally)

Reply:
> Does this not also affect query iteration?

No, it doesn't.  `Query` has a `&'s QueryState`, so we can copy that reference and hand it out with the full `'s` lifetime.  This change only affects methods that take `&mut QueryState`.  

(The issue with things like `QueryManyIter::fetch_next()` and `Query::iter_mut()` are that the `UnsafeWorldCell<'w>` is acting like a `&'w mut`, so we can't copy it and instead have to reborrow it with a shorter `'w` lifetime.)

Reply:
Okay, so this does present an issue within individual iterators.
However, this will be a breaking change to more methods of `QueryState` than what you've currently listed:
- `get`
- `get_single` (no checked manual version)
- `get_many` (no manual version)
- `single` (no manual version)
- `iter`
- `iter_many`
- `iter_combinations` (no manual version)
- `par_iter` (no manual version)
and naturally, any method that takes at least `&'a QueryState` while the query items returned by one of the above is live.

The lack of manual versions can be considered API holes for some of the above, however not all of them (f.e. `single`)

Do I understand correctly?

If so, I think the ability to mix these methods of `QueryState` contains important use cases, and I would see those as more valuable than this change in its current form :/